test/unit/org/apache/cassandra/cql3/validation/entities/CollectionsTest.java

```diff
@@ -2107,4 +2107,20 @@ public void testSetReversed() throws Throwable
         assertRows(execute("SELECT c['t3'..] FROM %s"), row(set("t3", "t4")));
         assertRows(execute("SELECT c[..'t5'] FROM %s"), row(set("t1", "t2", "t3", "t4")));
     }
+
+    /**
+     * {@link org.apache.cassandra.db.rows.Row.Builder} is cached, and there was logic to reuse the same builder cross
+     * rows, even if the builder was returned to the pool! This is "fine" in the happy path, but if the second row fails
+     * for any reason, then the builder is in a partial state and also referenced by the pool!
+     */
+    @Test
+    public void CASSANDRA_21055()
```

Comment:
i have confirmed if you revert the `src` changes this test fails as it detects the cache corruption.  `execute` runs in `main` so this test is deterministic and don't have to worry about multi threaded mutation stage.